test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -2818,55 +2801,31 @@ class test_ucp_sockaddr_protocols_err_sender
          * with CANCEL status */
         entity_disconnect(sender());
 
-        if (has_invalidate && (num_senders > 1)) {
-            for (auto &send_recv_info : send_recv_infos) {
-                /* Check that the send request would not be completed until the
-                 * memory handle for the same send buffer will not be released
-                 */
-                void *sreq          = std::get<0>(send_recv_info);
-                ucs_status_t status = request_progress(sreq, { &sender() },
-                                                       0.5);
-                ASSERT_EQ(UCS_INPROGRESS, status);
-            }
-        }
+        mem_buffer recv_buf(size, UCS_MEMORY_TYPE_HOST);
+        mem_buffer extra_recv_buf(size, UCS_MEMORY_TYPE_HOST);
 
-        auto first_send_recv_info = std::move(send_recv_infos.front());
-        send_recv_infos.pop_front();
-
-        for (auto &send_recv_info : send_recv_infos) {
-            auto message                = std::get<1>(send_recv_info);
-            std::get<2>(send_recv_info) = recv(receiver(), &extra_recv_buf[0],
-                                               size, message,
-                                               rtag_complete_cbx, NULL);
+        recv_buf.pattern_fill(2, size);
+        extra_recv_buf.pattern_fill(2, size);
+        for (size_t i = 1; i < messages.size(); ++i) {
+            void *rreq = recv(receiver(), extra_recv_buf.ptr(), size,
+                              messages[i], rtag_complete_always_ok_cbx, NULL);
+            reqs.push_back(rreq);
         }
 
-        while (!send_recv_infos.empty()) {
-            auto &send_recv_info = send_recv_infos.front();
-
-            status = requests_wait({ std::get<0>(send_recv_info),
-                                     std::get<2>(send_recv_info) });
-            ASSERT_UCS_OK(status);
-            send_recv_infos.pop_front();
-        }
-
-        void *sreq = std::get<0>(first_send_recv_info);
-        status     = request_wait(sreq);
+        status = requests_wait(reqs);
         ASSERT_EQ(UCS_ERR_CANCELED, status);
 
         /* Receive buffer should not be updated */
-        ASSERT_EQ(recv_buf, recv_copy);
+        recv_buf.pattern_check(2, size);
         /* Update send buffer by new data - emulation of free(buffer) */
-        memset(&send_buf[0], 'z', size);
+        send_buf.pattern_fill(3, size);
 
         /* Complete receive operation */
-        ucp_tag_message_h message = std::get<1>(first_send_recv_info);
-        void *rreq                = recv(receiver(), &recv_buf[0], size,
-                                         message, rtag_complete_cbx, NULL);
+        void *rreq = recv(receiver(), recv_buf.ptr(), size, messages[0],
+                          rtag_complete_check_data_cbx, recv_buf.ptr());
```

Comment:
can it be part of the loop in 2809? why we have to do it later?

Reply:
we have to do it later to fill the send buffer by another data